### PR TITLE
YJack0000-proposal-submission

### PR DESCRIPTION
**Pull request by YJack0000**
---
A good PR should include the following parts:

- [x] Provide a clear title
- [x] Provide meaningful descriptions in the PR and the commit logs
- [x] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [x] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [x] Exclude irrelevant files, such as binaries, text files, or dot files
- [x] Make sure tests/CI pass

### CODE REVIEW

File: project/YJack0000/README.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
+SimEvo: A Simple Python Library for Simulations of Natural Selection
+---
+
+## Basic Information
+
+> **Project Overview:** Simulating Natural Selection in Ecosystems using C++ and Python
+
+SimSel leverages the computational efficiency of C++ for the core simulation engine and offers an intuitive Python interface for ease of use and accessibility. It models ecosystems composed of organisms with genetically determined traits, interacting within an environment that includes resources like food. The simulation observes the principles of natural selection, with organism survival and reproduction influenced by their traits and environmental factors.
```
Comment (yungyuc) [inline project/YJack0000/README.md]: One paragraph is a line, and it makes the lines to be too long. It is hard to add comments at specific location for a long line. Please change to 79-char width.

File: project/YJack0000/README.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
+SimEvo: A Simple Python Library for Simulations of Natural Selection
```
Comment (yungyuc) [inline project/YJack0000/README.md]: I like the name

File: project/YJack0000/README.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+SimSel leverages the computational efficiency of C++ for the core simulation engine and offers an intuitive Python interface for ease of use and accessibility. It models ecosystems composed of organisms with genetically determined traits, interacting within an environment that includes resources like food. The simulation observes the principles of natural selection, with organism survival and reproduction influenced by their traits and environmental factors.
+
+- 📂 **GitHub Repository**: [https://github.com/YJack0000/SimEvo](https://github.com/YJack0000/SimEvo)
+- 💻 **Implemented In**:
+  - **Core Library**: C++
+  - **Bindings & Interface**: Python
+
+## Problem to Solve
+Understanding the dynamics of natural selection and ecosystem evolution involves complex interactions among organisms and their environments. SimSel aims to provide a simplified yet powerful simulation environment where users can define various organisms, their genetic makeup, and environmental factors, to observe how natural selection unfolds over generations.
```
Comment (yungyuc) [inline project/YJack0000/README.md]: SimEvo, or SimSel?

And now you see how the long lines make it hard to pin-point the target of the comment.  Fix it, please.

File: project/YJack0000/README.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+In this context, the implementation of spatial indexing algorithms plays a critical role. Efficient spatial indexing is essential for managing the simulation's computational complexity, enabling the simulation to handle large numbers of organisms and their interactions within the environment accurately and swiftly.
```
Comment (yungyuc) [inline project/YJack0000/README.md]: Please be specific to the spatial index that you plan to develop.

File: project/YJack0000/README.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+- **Educators and Students**: For teaching and learning evolutionary concepts through interactive simulations.
+- **Researchers**: For conducting experiments on evolutionary strategies and hypotheses.
+- **Hobbyists**: For exploring the principles of evolution in a virtual environment.
+
+## System Architecture
+SimSel comprises a high-performance simulation engine written in C++ for handling the complex computations of natural selection dynamics efficiently. Python bindings, created using pybind11, make this engine accessible to users who can interact with the simulation using a simple Python API.
+
+### Components:
+
+* **Genetic Traits and Organism Definition (C++)**: Models organisms with specific genetic traits that affect their survival and reproduction.
```
Comment (yungyuc) [inline project/YJack0000/README.md]: Genetic information, environmental models, and simulation engines mean a lot of code.  They are no less difficult than implementing a spatial index.  What is the focus of this project?  You seem to have 4, which are a lot.

File: project/YJack0000/README.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    - Markdown
+- **Continuous Integration**
+  - [Github Actions](https://github.com/features/actions)
+
+## Schedule
+
+Week 1 (4/8):
+- Set up and model the basic C++ environment.
+- Studying the relevant field knowledge and adjust the existing interface accordingly.
+- Set up a CI pipeline for unit testing.
```
Comment (yungyuc) [inline project/YJack0000/README.md:203]: It is very wise to set up CI in the first place.  You won't regret.

File: project/YJack0000/README.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+**ISpatialObject Interface**: Represents any entity within the simulation environment that occupies space, such as organisms and food items. This interface ensures that the spatial index can manage a variety of object types, making the simulation more versatile.
+
+### Default and Optimized Implementations
+
+**Default Implementation**: In the initial stages or in simulations where the density of entities is low, a brute-force approach is employed. This method involves calculating the distances between all pairs of objects to determine proximity-based interactions. While straightforward, this approach is computationally intensive as the number of entities increases.
+
+**Optimized Implementation**: For more complex simulations with a higher density of entities, SimSel utilizes a Quadtree data structure for spatial indexing. Quadtrees are highly efficient for 2D spatial data, allowing for rapid queries of objects within specific areas. This optimization significantly reduces the computational cost of distance calculations and interaction determinations, making the simulation more scalable and performant.
```
Comment (yungyuc) [inline project/YJack0000/README.md]: If quadtree is your focus of the project, you are mentioning it too late.  It makes me wonder maybe you are focusing on something else, and baffles me.

You need to make your goal and focus clearer.

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] I can't find your submission on E3.  Please add it there.
- [x] You missed an entry in `project/README.rst` as [requested](https://github.com/yungyuc/nsdhw_24sp/blob/567be93bc63a6fb1c2c55b8faf5825172daeaba5/project/README.rst?plain=1#L22)
- [x] 79-char line width
- [x] Specify the spatial index to be developed
- [ ] Specify the focus of the project among genetic information, environmental models, simulation engines, and spatial index

When you are ready for the next review, leave a comment, clearly write down something like "I am ready, please review again".

File: project/YJack0000/README.md (hunk 3 above)
Comment (YJack0000) [reply project/YJack0000/README.md]: A big thanks for taking the time to go through everything in detail! Fixing it now! 
 The reason for the inconsistency is that I was alternating between using 'SimEvo' and 'SimSel' while drafting this proposal. 🤣

File: project/YJack0000/README.md (hunk 1 above)
Comment (YJack0000) [reply project/YJack0000/README.md]: [fix: 79-char width](https://github.com/YJack0000/nsdhw_24sp/commit/f44041224abb2d016d984d7ce6d8400a7fe94dc1)

File: project/YJack0000/README.md (hunk 5 above)
Comment (YJack0000) [reply project/YJack0000/README.md]: I'm really aiming to explore a specific problem in this kind of simulation in 2d space, with Quadtree indexing being a practical choice to improve efficiency. Planning to keep other aspects simpler to maintain focus.

File: project/YJack0000/README.md (hunk 4 above)
Comment (YJack0000) [reply project/YJack0000/README.md]: [fix: mention quadtree on problem to solve](https://github.com/YJack0000/nsdhw_24sp/commit/343376ccb150a1a21aacaa22d59d8b938888c112)

**Comment by YJack0000** [comment]
@yungyuc I've addressed all the issues. Please review again. Thanks.

**Comment by YJack0000** [comment]
> @yungyuc I've addressed all the issues. Please review again. Thanks.



> @yungyuc I've addressed all the issues. Please review again. Thanks.

213f6907b9839ef2858d7df64318abb86710a97b

File: project/YJack0000/README.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+- 📂 **GitHub Repository**:
+  [https://github.com/YJack0000/SimEvo](https://github.com/YJack0000/SimEvo)
+- 💻 **Implemented In**:
+  - **Core Library**: C++
+  - **Bindings & Interface**: Python
+
+## Problem to Solve
+
+Understanding the dynamics of natural selection and ecosystem evolution
```
Comment (YJack0000) [inline project/YJack0000/README.md:24]: fix: mention quadtree on problem to solve
https://github.com/YJack0000/nsdhw_24sp/commit/343376ccb150a1a21aacaa22d59d8b938888c112

File: project/YJack0000/README.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+involves complex interactions among organisms and their environments, where
+spatial dynamics are crucial. The core challenge SimEvo addresses is
+efficiently managing these interactions through advanced spatial indexing,
+specifically using Quadtree algorithms. This approach is fundamental for
+accurately and swiftly handling the complex web of organism interactions and
+environmental factors, essential for observing natural selection's unfolding
+over generations.
```
Comment (YJack0000) [inline project/YJack0000/README.md:31]: fix: mention quadtree on problem to solve
https://github.com/YJack0000/nsdhw_24sp/commit/343376ccb150a1a21aacaa22d59d8b938888c112

File: project/YJack0000/README.md (hunk 7 above)
Comment (YJack0000) [reply project/YJack0000/README.md]: [fix: mention quadtree on problem to solve](https://github.com/YJack0000/nsdhw_24sp/commit/343376ccb150a1a21aacaa22d59d8b938888c112)

Is this early mention of Quadtree in the 'Problem to Solve' section sufficient, or should I add more detailed context about Quadtree?

File: project/YJack0000/README.md (hunk 3 above)
Comment (yungyuc) [reply project/YJack0000/README.md]: It's not hard to do a replace-all.  But it's OK.  I am fine with minor flaws like this.  Having said that, it's always good to be rigorous in your edit.

File: project/YJack0000/README.md (hunk 5 above)
Comment (yungyuc) [reply project/YJack0000/README.md]: This gonna be challenging.  Make sure you are flexible about the outcome.  Talking to more experienced developers will help you a lot.  If you really want to hit the target, sciwork meetups and sprints will help you a lot.

File: project/YJack0000/README.md (hunk 7 above)
Comment (yungyuc) [reply project/YJack0000/README.md]: Since it's not the focus of the project, it's fine to tone down it.

**Review by yungyuc — APPROVED** [review]
I don't think you have made clear the focus of the project in the proposal, while providing the ideas in your comments, but it's fine.  You will need very good management of your time.

**Pull request merged**